crates/cranelift/src/func_environ.rs

```diff
@@ -19,109 +20,74 @@ use cranelift_wasm::{
+/// array so it can be known which libcalls are needed for this
+/// function.
+struct BuiltinFunctions {
+    types: BuiltinFunctionSignatures,
 
-            #[cfg(feature = "gc")]
-            reference_type: ir::Type,
+    builtins:
+        [Option<ir::FuncRef>; BuiltinFunctionIndex::builtin_functions_total_number() as usize],
 
-            call_conv: isa::CallConv,
+    required: Vec<BuiltinFunctionIndex>,
+}
 
-            $(
-                $( #[$attr] )*
-                $name: Option<ir::SigRef>,
-            )*
+impl BuiltinFunctions {
+    fn new(isa: &dyn TargetIsa) -> Self {
+        Self {
+            types: BuiltinFunctionSignatures::new(isa),
+            builtins: [None; BuiltinFunctionIndex::builtin_functions_total_number() as usize],
+            required: Vec::new(),
         }
+    }
 
-        #[allow(unused_doc_comments)]
-        impl BuiltinFunctionSignatures {
-            fn new(
-                pointer_type: ir::Type,
-                reference_type: ir::Type,
-                call_conv: isa::CallConv,
-            ) -> Self {
-                #[cfg(not(feature = "gc"))]
-                let _ = reference_type;
-
-                Self {
-                    pointer_type,
-
-                    #[cfg(feature = "gc")]
-                    reference_type,
-
-                    call_conv,
-
-                    $(
-                        $( #[$attr] )*
-                        $name: None,
-                    )*
-                }
-            }
-
-            fn vmctx(&self) -> AbiParam {
-                AbiParam::special(self.pointer_type, ArgumentPurpose::VMContext)
-            }
-
-            #[cfg(feature = "gc")]
-            fn reference(&self) -> AbiParam {
-                AbiParam::new(self.reference_type)
-            }
-
-            fn pointer(&self) -> AbiParam {
-                AbiParam::new(self.pointer_type)
-            }
-
-            fn i32(&self) -> AbiParam {
-                // Some platform ABIs require i32 values to be zero- or sign-
-                // extended to the full register width.  We need to indicate
-                // this here by using the appropriate .uext or .sext attribute.
-                // The attribute can be added unconditionally; platforms whose
-                // ABI does not require such extensions will simply ignore it.
-                // Note that currently all i32 arguments or return values used
-                // by builtin functions are unsigned, so we always use .uext.
-                // If that ever changes, we will have to add a second type
-                // marker here.
-                AbiParam::new(I32).uext()
-            }
+    fn load_builtin(&mut self, func: &mut Function, index: BuiltinFunctionIndex) -> ir::FuncRef {
+        let cache = &mut self.builtins[index.index() as usize];
+        if let Some(f) = cache {
+            return *f;
+        }
+        self.required.push(index);
```

Comment:
Should required be a set of some sort?

Reply:
I think I'll actually refactor and remove this since I realized it's also present via relocations, so I'll handle this when I rebase over the winch trampolines going through Cranelift.